### PR TITLE
Check database version in migrate command

### PR DESCRIPTION
**Pull request by ausi**
As discussed in https://github.com/contao/contao/pull/4158#issuecomment-1169685620 (and on Slack)

This only fails if the configured server version causes the wrong doctrine platform to be used. 
So if the actual version is `8.0.29` but `8.0.28` is configured no error occurs.
But if `5.7.0` is configured it would show an error message like: 
`Wrong database version, please set it to "8.0.29". Expected "Doctrine\DBAL\Platforms\MySQL80Platform" was "Doctrine\DBAL\Platforms\MySQLPlatform".`

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]
Tested and looks good to me. (However, I don't find the _"Expected ..."_ part of the error message particularly useful.)

**Comment by fritzmg** [comment]
Btw. shouldn't this go into `4.9`? It would be just as helpful there (as for performance you can also define the database version in your `parameters.yaml`, `DATABASE_URL` environment variable or `config.yaml` there).

**Comment by ausi** [comment]
> Btw. shouldn't this go into `4.9`?

Because of several merge conflicts that would arise I created a separate PR for 4.9 now: #5101

**Review by aschempp — CHANGES_REQUESTED** [review]
If the output format is `ndjson`, this should not throw an exception (result in a `error` type message) but generate a `problem` (see https://github.com/contao/contao/pull/5065). Otherwise the Contao Manager cannot see if there is a connection problem/runtime exception or a database-level error.

**Comment by ausi** [comment]
> If the output format is `ndjson`, this should not throw an exception

Done in 6340b8642a842f653645fd7fc166be1d9fe6400e

File: core-bundle/src/Command/MigrateCommand.php [hunk 1]
```diff
@@ -537,4 +546,40 @@ private function compileTableWarnings(Table $table): array
...
+        }
+
+        $message = sprintf('Wrong database version configured, please set it to "%s"', $version);
+
+        if ($currentVersion = $this->connection->getParams()['serverVersion'] ?? null) {
+            $message .= sprintf(', currently set to "%s"', $currentVersion);
+        }
+
+        if ($asJson) {
+            $this->writeNdjson('database-error', ['message' => $message]);
```
Comment (ausi) [inline core-bundle/src/Command/MigrateCommand.php]: The correct name needs to be resolved in https://github.com/contao/contao/pull/5065#discussion_r940079950 first

Comment (ausi) [reply core-bundle/src/Command/MigrateCommand.php]:
```suggestion
            $this->writeNdjson('problem', ['message' => $message]);
```

File: core-bundle/tests/Command/MigrateCommandTest.php [hunk 2]
```diff
@@ -429,6 +434,65 @@ public function testDoesAbortOnFatalError(string $format): void
...
+
+        $code = $tester->execute(['--format' => $format, '--no-backup' => true], ['interactive' => 'ndjson' !== $format]);
+        $display = $tester->getDisplay();
+
+        $this->assertSame(1, $code);
+
+        if ('ndjson' === $format) {
+            $json = $this->jsonArrayFromNdjson($display)[0];
+
+            $this->assertSame('database-error', $json['type']);
```
Comment (ausi) [inline core-bundle/tests/Command/MigrateCommandTest.php]:
```suggestion
            $this->assertSame('problem', $json['type']);
```

**Review by aschempp — DISMISSED** [review]
not sure why we need this for upstream merge, but code looks good 🙃

File: core-bundle/src/Command/MigrateCommand.php [hunk 3]
```diff
@@ -537,4 +546,40 @@ private function compileTableWarnings(Table $table): array
 
         return $warnings;
     }
+
+    private function validateDatabaseVersion(bool $asJson): bool
+    {
+        // TODO: Find a replacement for getWrappedConnection() once doctrine/dbal 4.0 is released
+        $driverConnection = $this->connection->getWrappedConnection();
+        $currentPlatform = \get_class($this->connection->getDatabasePlatform());
```
Comment (leofeyer) [inline core-bundle/src/Command/MigrateCommand.php]: Is this going to be a problem?

Comment (ausi) [reply core-bundle/src/Command/MigrateCommand.php]: I don’t think so, see https://github.com/contao/contao/pull/5133

Comment (leofeyer) [reply core-bundle/src/Command/MigrateCommand.php]: See https://github.com/ausi/contao-contao/pull/4 😄

Comment (ausi) [reply core-bundle/src/Command/MigrateCommand.php]: Merged.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**